### DISCUSSION TITLE
Week 2 Check Ins

### DISCUSSION THREAD

**Discussion by minhphan03** [UW-Upwelling-Project/Summer-2023#6] [General]
[Placeholder]

**Comment by NaNa7Miiii** [comment]
### Jiarui's Tuesday Update
According to Dr.Eli's method of establishing buffer areas and line-smoothing, I searched for ways of how similar ways can be conducted via Python. 
# Import required packages

import matplotlib.pyplot as plt
import geopandas as gpd
import cartopy.crs as ccrs
from shapely.geometry import Polygon, MultiPolygon
from shapely.ops import unary_union
import numpy as np
from matplotlib.patches import Polygon as mplPolygon
import matplotlib.ticker as mticker

# Select the India coastline from the world map
'crs = ccrs.PlateCarree()

coastline = gpd.read_file(gpd.datasets.get_path('naturalearth_lowres'))

india_coastline = coastline[coastline['name'] == 'India']

# Convert the CRS to UTM zone 43N
india_coastline_utm = india_coastline.to_crs('EPSG:32643')

# Creating 100km and 300km buffer lines along coastline
distances = [100000, 300000]

buffered = []

for distance in distances:
    india_buffer = india_coastline_utm.geometry.apply(lambda x: x.buffer(distance))
    buffered.append(india_buffer)

buffered_wgs84 = [buf.to_crs(india_coastline.crs) for buf in buffered]

# Unify all buffered geometries
union_buffered = unary_union(buffered)

# Plotting
fig = plt.figure()
ax = plt.axes(projection=ccrs.PlateCarree())
ax.set_extent([lons.start, lons.stop, lats.start, lats.stop], crs=ccrs.PlateCarree())
ax.coastlines(resolution='10m', color='black', linewidth=1)

# Add gridlines
gl = ax.gridlines(crs=ccrs.PlateCarree(), draw_labels=True, linewidth=1, color='gray', alpha=0.5, linestyle='--')
gl.xlabels_top = False
gl.ylabels_left = False
gl.xlines = True
gl.xlocator = mticker.FixedLocator(range(int(lons.start), int(lons.stop)+1, 5))   # Set longitude markers every 5 degrees
gl.ylocator = mticker.FixedLocator(range(int(lats.start), int(lats.stop)+1, 5))   # Set latitude markers every 5 degrees

# Plot buffers
for buf in buffered_wgs84:
    for geom in buf.geometry:
        if geom.is_empty:
            continue
        if geom.geom_type == 'Polygon':
            geom = MultiPolygon([geom])
        for polygon in geom.geoms:
            # Create the matplotlib Polygon and add it to the plot
            x, y = polygon.exterior.coords.xy
            coords = np.column_stack((x, y))
            mpl_poly = mplPolygon(coords, edgecolor='red', facecolor='none', alpha=0.5)
            ax.add_patch(mpl_poly)

plt.show()`

**This is the output of applying a 100km and a 300km buffer zone along the coastal line (there is noise on the top):**
 
<img width="463" alt="Screenshot 2023-06-20 at 3 55 03 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/768e6317-b9bc-43e8-97c0-13ac662eef46">

Then I tried to detect upwelling via the sst differences, this is my notebook: https://dhub.opensci.live/user/nana7miiii/lab/tree/test.ipynb

**Comment by minhphan03** [comment]
# Minh's Monday check-in:
ERA5 data: downloaded northward and eastward wind data, cleaned and combined data.
Took average of chlorophyll data into csv file
Surface currents: only downloaded monthly data since there was a lot to handle: will try to subset and combine data (daily, if possible) later.
Still needs to hear about hawaii-based institute published wind speed source

**Comment by NaNa7Miiii** [comment]
@eeholmes I have used ds.attrs to check the metadata to find the projection, but there is no projection info (I am using links such as 'https://archive.podaac.earthdata.nasa.gov/podaac-ops-cumulus-protected/MUR-JPL-L4-GLOB-v4.1/20210701090000-JPL-L4_GHRSST-SSTfnd-MUR-GLOB-v02.0-fv04.1.nc'). There are some attributes such as 
    **'geospatial_lat_units': 'degrees north'
    'geospatial_lon_units': 'degrees east'**
indicate that the dataset's coordinate units are degrees (specifically, north latitude and east longitude), which usually means that the data uses a geographic coordinate system, specifically the WGS 84 coordinate system. Yesterday we discussed the possibility that **crs = ccrs.PlateCarree()** may use a different projection from **buffered_wgs84 = [buf.to_crs(india_coastline.crs) for buf in buffered]**, while I just checked that the code for **crs** is **EPSG:4326**, and using **buffered_plate_carree = [buf.to_crs('EPSG:4326') for buf in buffered]** plots the same diagram compared with using buffered_wgs84, indicating **the projection is the same**.

**Reply by NaNa7Miiii** [reply]
Maybe it's my buffer setup problem. Also, I noticed my algorithm compares the sst difference for points in two buffers at the same latitude, which can be acceptable for horizontal cases, but for cases at the bottom of India, the buffers cannot use horizontal comparison, so I will find out a way to adjust my algorithm.

**Reply by NaNa7Miiii** [reply]
### Jiarui's Wed Update:
My previous code of finding nearest point for every lat is problematic: 
`def find_nearest_point(lat, buffer_geom):
    point_on_equator = Point(0, lat)
    nearest_point = nearest_points(buffer_geom, point_on_equator)[0]
    return nearest_point`
I am using a new function:
<img width="953" alt="Screenshot 2023-06-21 at 3 53 53 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/70557338-16ad-4638-b3e4-0a1cafaf4301">
to get the points. The results is better:
<img width="635" alt="Screenshot 2023-06-21 at 3 54 47 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/2fe46fba-e2c0-440e-bbee-d7ecfbd62ac3">
This result seems can be a solution for solving the noise where the buffer zone curved into the land:
<img width="511" alt="Screenshot 2023-06-21 at 10 19 10 AM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/e7d680fb-4402-46e6-880e-fc5f66ad57fe">
Of course, one can always customize how many points that need to be calculated. For example, 
<img width="602" alt="Screenshot 2023-06-21 at 3 22 34 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/c3a6e84f-3e47-475e-b0fb-e6696a3fb19c">
My new func still has some problems, but hopefully, I will get them fixed by tomorrow!

**Comment by minhphan03** [comment]
# Minh's Wednesday check-in:
Finish downsampling era5 wind data, computed speed and direction.
Get India's border coordinates
Explore datasets: wave direction, surface current
Upload files to GitHub.

**Reply by eeholmes** [reply]
@minhphan03 @NaNa7Miiii Here is an example of using tensorflow with satellite raster images stored as geotiff files https://towardsdatascience.com/neural-network-for-satellite-data-classification-using-tensorflow-in-python-a13bcf38f3e1

**Reply by eeholmes** [reply]
And I think that each geoTiff image can have multiple layers (sst, ssh, wind-speed)

**Comment by eeholmes** [comment]
# Eli's Thursday Check-in

I talked with colleagues about how to store data. A Zarr database might be a good option. https://zarr.dev/slides/scipy-2019.pdf, https://zarr.readthedocs.io/en/stable/. It sounds like this is an efficient way to work with data that are too big for memory and is optimized for parallel computing.

Other notes
* blog on netCDF and Tensorflow https://www.noahbrenowitz.com/post/loading_netcdfs/
* images to Zarr https://forum.image.sc/t/manually-creating-ome-zarr-from-a-folder-of-tiffs-pngs/59806
* OME-Zarr might be helpful https://www.biorxiv.org/content/10.1101/2023.02.17.528834v1.full.pdf
* Instructions on this image -> Zarr -> tensorflow pipeline: https://meetingorganizer.copernicus.org/EGU22/EGU22-13193.html
* https://github.com/pangeo-data/pangeo-datastore
* https://climetlab.readthedocs.io

I have a meeting set up for Friday to get the storage set up. Need to get an idea of the size of storage I should set up
*

**Comment by NaNa7Miiii** [comment]
### Jiarui's Thursday Check-in
The method of creating buffers I previously used was to create buffers for the entire Indian coastal line, so today I clipped the buffers to our area of interest.
I have made the function to find the closest point on the 300km buffer related to the 100km buffer point:

<img width="1003" alt="Screenshot 2023-06-22 at 3 54 06 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/dd37284f-fb02-4441-9307-265956242b4d">
<img width="507" alt="Screenshot 2023-06-22 at 3 54 47 PM" src="https://github.com/UW-Upwelling-Project/Summer-2023/assets/97627889/bc3520b0-fa39-4cc5-9a8c-3b3008100571">

There are still some mismatch bugs to be fixed.

**Comment by eeholmes** [comment]
Eli Check-in

@minhphan03 Some other sources of ERA5 that might allow applying a bounding box before downloading

* https://developers.google.com/earth-engine/datasets/catalog/ECMWF_ERA5_DAILY
* Doesn't actually look like applying a bounding box is possible but I don't know google earth engine

Alternative approach. Go back to using ERDDAP where you can apply a bounding box.
http://apdrc.soest.hawaii.edu/erddap/griddap/hawaii_soest_aa5a_f355_2d1c.html

1. Set bounding box lat 0,30 lon 50, 100 (so bigger to get all of indian ocean and bay of bengal)
2. Download nc file for a day using a URL (just change the date bit in your url construction)
http://apdrc.soest.hawaii.edu/erddap/griddap/hawaii_soest_aa5a_f355_2d1c.nc?u10[(2023-05-31T00:00:00Z):1:(2023-05-31T23:00:00Z)][(0.0):1:(30.0)][(50.0):1:(100.0)]
4. Convert nc file to xarray and average over 24hr to create daily 10m winds
5. Save that as nc file with title like 2020-05-01-10m-u-wind.nc
6. Delete original hourly file
7. Repeat for every day and all variables (10m u wind, 10m v wind, sst, surface total precipitation, daily 3d Geopotential). Those are the 5 of interest in ERDDAP

Link discussing ERA5
* https://rmets.onlinelibrary.wiley.com/doi/full/10.1002/qj.3803
* discussing 10m vs 100m winds https://confluence.ecmwf.int/pages/viewpage.action?pageId=250024041

Geopotential and upwelling?
* https://core.ac.uk/download/pdf/33016988.pdf
* http://www.jeeng.net/Propagation-of-Upwelling-on-Western-Coast-Sumatera-During-MJO-Event,79420,0,2.html
* https://journals.ametsoc.org/view/journals/phoc/47/1/jpo-d-16-0176.1.xml
* https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2009GL042231

**Reply by minhphan03** [reply]
I'm thinking maybe instead of averaging wind data per day, we talked about how night and day winds will be drastically different and maybe I can take samples of wind data at, say, every 12 hours (12am and 12pm). The ERDDAP dataset have strides that I can customize.

**Reply by minhphan03** [reply]
## Some key ideas from the sources you provided for the correlation between geopotential and upwelling: (pardon for the improper citations)
* A combination of the tropical depression and the NPH contributed to the generation of the strong and consistent southerly winds, eventually resulting in the Ekman-induced coastal upwelling. (Park, K.-A., and Kim, K.-R. (2010),) **Southerly winds to the west coast of India acting as shores to the left of the wind direction, which is consistent to the concept of coastal upwelling. By Coriolis effect, the southern winds would push the warmer water currents to the west in the summer and allow colder water to emerge**
* Upwelling during summer should occur in the deep waters and the upward currents should cease to exist through the core of thermocline which becomes more stratified disallowing upward movement of water. (P. N. LATHIPHA & A. V. S. MURTY) **Need to investigate water pressure at different ocean levels?**
* A mixing at the top layers reduces the core of thermocline and thereby intensifies the process of vertical transfer of properties **reduce the core by depth, I guess?**
* Heat and salt are, therefore, to be transferred from the depth where upwelling ceases to the Subsurface and surface by means of eddy diffusion. **upwelling only occurs above a certain depth (?)  // Investigate temperature at certain sea depth levels?**
* The upwelling process itself is a process whereby a sea mass is pushed upward along the continent, when the beach is to the left of the wind direction, the ecological transport leads to the mass of water away from the coast. As a result, there is a mass vacuum (divergence) in the coastal area. This mass void will be filled by the mass of water from the inner layer that moves to the surface. (Haryanto, Y. D. (2018)) **another upwelling process definition**

**Reply by minhphan03** [reply]
Also the Copernicus database has geopotential reanalysis. I'll download some data and let you know what I found...

Geopotential data has a level parameter (vertical to x and y), with levels from 100m to 1000m for the ERDDAP dataset I found ([link](http://apdrc.soest.hawaii.edu/erddap/griddap/hawaii_soest_1d9a_1b67_afa4.html)) Recommend to specify a range we're interested in to speed up downloading processes. 

Edit: it seems like Lathipha & Murity's paper aims to record the surface geopotential only: "The surface geopotential thus obtained for each season is considered in the analysis" (P. N. LATHIPHA & A. V. S. MURTY).
